### PR TITLE
Support GCE alpha/beta api endpoint override

### LINKED ISSUE #48568: GCE: ApiEndpoint specifies alpha/beta/v1 but doesn't support simultaneous use
#47881 breaks several things

1) The api endpoint is hardcoded in multiple places of the cloud provider.
2) Api endpoint is only used for main client, not the beta client which is used simultaneously. 
2) The api endpoint flag specifies alpha/beta/v1; however, we have two GCE clients (v1 and beta). How does that work?

/assign @cadmuxe

### PR DESCRIPTION
**Pull request by freehan**
fixes: https://github.com/kubernetes/kubernetes/issues/48568

### CODE REVIEW

**Comment by nicksardo** [comment]
There are many hardcoded references to `https://googleapis.com/...` in other GCE files. Can we tackle those here too? Probably best to use `gce.service.BasePath` in those cases.

File: pkg/cloudprovider/providers/gce/gce.go [hunk 1]
```diff
@@ -248,16 +253,24 @@ func CreateGCECloud(apiEndpoint, projectID, region, zone string, managedZones []
...
-	serviceBeta, err := computebeta.New(client)
+	serviceAlpha, err := computealpha.New(client)
 	if err != nil {
 		return nil, err
 	}
 
+	if apiEndpoint != "" {
+		service.BasePath = fmt.Sprintf("%sprojects/", apiEndpoint)
+		serviceBeta.BasePath = fmt.Sprintf("%sprojects/", strings.Replace(apiEndpoint, "v1", "beta", 0))
+		serviceAlpha.BasePath = fmt.Sprintf("%sprojects/", strings.Replace(apiEndpoint, "v1", "alpha", 0))
```
Comment (nicksardo) [inline pkg/cloudprovider/providers/gce/gce.go]: Do we document on the flag that `v1` is expected?

Comment (freehan) [inline pkg/cloudprovider/providers/gce/gce.go]: It seems like it:
prod: https://www.googleapis.com/compute/v1/
stage: https://www.googleapis.com/compute/staging_v1/
devcluster: http://localhost:3990/compute/v1

Comment (freehan) [inline pkg/cloudprovider/providers/gce/gce.go]: I will add a comment.

**Comment by freehan** [comment]
rebased

**Comment by freehan** [comment]
fixed unit test

**Comment by nicksardo** [comment]
Could you rename something in this PR that's related? 
https://github.com/kubernetes/kubernetes/blob/0744964956abdb17586361dee59853f6167b58d7/pkg/cloudprovider/providers/gce/gce_instances.go#L28 
This package is aliased `computealpha` but refers to the beta api.

File: pkg/cloudprovider/providers/gce/gce.go (hunk 1 above)
Comment (nicksardo) [inline pkg/cloudprovider/providers/gce/gce.go]: I was suggesting adding to the user-viewable side: https://github.com/kubernetes/kubernetes/pull/47881/files#diff-e00287b5331f6d39959423afadb6b368R207.

**Comment by nicksardo** [comment]
Awesome.
/lgtm
/approve

File: pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go [hunk 2]
```diff
@@ -671,10 +672,9 @@ func nodeNames(nodes []*v1.Node) []string {
 	return ret
 }
 
-func makeHostURL(projectID, zone, host string) string {
+func makeHostURL(apiEndpiont, projectID, zone, host string) string {
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go]: spelling of apiEndpoint

File: pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go [hunk 3]
```diff
@@ -624,7 +625,7 @@ func getPortsAndProtocol(svcPorts []v1.ServicePort) (ports []string, protocol v1
 }
 
 func (gce *GCECloud) getBackendServiceLink(name string) string {
-	return fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/regions/%s/backendServices/%s", gce.projectID, gce.region, name)
+	return gce.service.BasePath + path.Join("projects", gce.projectID, "regions", gce.region, "backendServices", name)
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go]: probably should use strings.Join

File: cluster/gce/config-default.sh [hunk 4]
```diff
@@ -72,12 +75,12 @@ GCI_VERSION=${KUBE_GCI_VERSION:-cos-stable-59-9460-64-0}
 MASTER_IMAGE=${KUBE_GCE_MASTER_IMAGE:-}
 MASTER_IMAGE_PROJECT=${KUBE_GCE_MASTER_PROJECT:-cos-cloud}
 NODE_IMAGE=${KUBE_GCE_NODE_IMAGE:-${CVM_VERSION}}
-GCE_API_ENDPOINT=${KUBE_GCE_API_ENDPOINT:-}
 NODE_IMAGE_PROJECT=${KUBE_GCE_NODE_PROJECT:-cos-cloud}
 CONTAINER_RUNTIME=${KUBE_CONTAINER_RUNTIME:-docker}
 RKT_VERSION=${KUBE_RKT_VERSION:-1.23.0}
 RKT_STAGE1_IMAGE=${KUBE_RKT_STAGE1_IMAGE:-coreos.com/rkt/stage1-coreos}
 
+
```
Comment (bowei) [inline cluster/gce/config-default.sh]: delete

File: pkg/cloudprovider/providers/gce/gce.go [hunk 5]
```diff
@@ -273,16 +279,28 @@ func CreateGCECloud(apiEndpoint, projectID, networkProjectID, region, zone strin
 		return nil, err
 	}
 
-	if apiEndpoint != "" {
-		service.BasePath = fmt.Sprintf("%sprojects/", apiEndpoint)
+	client, err = newOauthClient(tokenSource)
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce.go]: `err` seems to be not checked here

File: pkg/cloudprovider/providers/gce/gce.go [hunk 6]
```diff
@@ -273,16 +279,28 @@ func CreateGCECloud(apiEndpoint, projectID, networkProjectID, region, zone strin
...
+	serviceBeta, err := computebeta.New(client)
+	if err != nil {
+		return nil, err
 	}
 
 	client, err = newOauthClient(tokenSource)
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce.go]: `err` seems to be not checked here

File: pkg/cloudprovider/providers/gce/gce.go (hunk 5 above)
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce.go]: fix the instance above too

**Comment by freehan** [comment]
fixed comments

**Review by bowei — COMMENTED** [review]
/lgtm

**Comment by freehan** [comment]
@thockin for /approve

**Comment by bowei** [comment]
@wojtek-t for approval

**Comment by freehan** [comment]
fixed tests...

**Comment by freehan** [comment]
Rebased the 2nd time

apply lgtm and bump priority

**Pull request merged**